[quant][graphmode] insert prepack/unpack for aten::linear

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #34518 [quant][graphmode] quantization work for prim::If
* #34517 [quant][graphmode][refactor] separate code for handling subblock in CallMethod
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* #34348 [quant] Add dequantize.tensors
* #34347 [quant][graphmode] Add quantization support for aten::dropout
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34315 [quant][graphmode] insert quant/dequant work for duplicated debugName
* #34314 [quant][graphmode][refactor] Change QParamMap to QParamVector
* **#33895 [quant][graphmode] insert prepack/unpack for aten::linear**
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops

Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_jit.py:
```diff
@@ -2052,6 +2052,22 @@ def forward(self, x, weight, bias):
         res = m(x, weight, bias)
         self.assertEqual(res, ref_res)
 
+    def test_finalize(self):
```
Please move test to #33927

jerryzh168 replied:
this test is only possible after this PR, but I can add another test there that tests conv2d

raghuramank100 commented on test/test_jit.py:
```diff
@@ -2052,6 +2052,22 @@ def forward(self, x, weight, bias):
+        model = quantize_script(model, qconfig_dict, _test_only_eval_fn, [data], inplace=False)
+        FileCheck().check("quantized::linear") \
+                   .run(model.graph)
+
```
Please add a test to verify that pre_pack op is present in the graph so that the pattern match is tested.

jerryzh168 replied:
quantized::linear is looking for prerpack ops, so when we see quantized::linear in the graph we can be sure that prepacked ops are there

raghuramank100 reviewed (changes requested):
Please see comments, thanks

jerryzh168 committed:
Update on "[quant][graphmode] insert prepack/unpack for aten::linear"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
deprecated by https://github.com/pytorch/pytorch/pull/33854
